### DISCUSSION TITLE
単体テストの考え方/使い方 9章

### DISCUSSION THREAD

**Discussion by y-kaneda** [atman-inc/training#104] [General]
輪読会のサマリ用ディスカッション。
9章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by y-kaneda** [comment]
# 第9章 モックのベスト・プラクティス

**Comment by y-kaneda** [comment]
## 9.1 モックの価値を最大限に引き出す方法

**Reply by nmurakami0** [reply]
- ここで出てくるプロセス外依存
  - DB
  - ログ出力オブジェクト (サポートログ)
  - メッセージバス
- DBは実装の詳細で、残り２つは観測可能な振る舞いの一部なので、モックしてテストする

**Comment by y-kaneda** [comment]
### 9.1.1 アプリケーションの境界を超えて行われるコミュニケーションの検証

**Reply by nmurakami0** [reply]
- IMessageBusだとシステムの境界よりも内側に入りすぎている
  - 実際に外部(メッセージバス)とのやり取りをするのはIBusのため
- IBusをモックすると、実際に送るテキストメッセージ自体を確認することが出来る
  - IMessageBusをモックするとメソッド呼び出しの確認になり、振る舞いより実装よりになってしまっている
    - その分regressionへの保護と、リファクタリングへの耐性が弱まってしまう


1クラス1責務になっていないとモックするのが大変になるので、IBusがシンプルになっているのが効いていそう

**Comment by y-kaneda** [comment]
### 9.1.2 モックからスパイへの置き換え

**Comment by y-kaneda** [comment]
### 9.1.3 ログ出力の検証でもテスト・ダブルの対象を変えるべきか？

**Comment by y-kaneda** [comment]
## 9.2 モックのベスト・プラクティス

**Comment by y-kaneda** [comment]
### 9.2.1 モックの利用は統合テストに限定する

**Comment by y-kaneda** [comment]
### 9.2.2 1つのテスト・ケースには複数のモックを持たせてはならない、という誤解

**Comment by y-kaneda** [comment]
### 9.2.3 モックの呼び出し回数を常に確認する

**Reply by nmurakami0** [reply]
回数が変わると振る舞いが変わってしまう

**Comment by y-kaneda** [comment]
### 9.2.4 モックの対象になる型は自身のプロジェクトが所有する型のみにする

**Reply by nmurakami0** [reply]
rspecのようにtime travelが用意されていない場合に、時間系のクラスをモックしてしまうときがあるが、
adapterを用意してそちらをモックすべき？

**Comment by y-kaneda** [comment]
## まとめ
